### PR TITLE
Much better unchecked warnings.

### PR DESCRIPTION
**Pull request by paulp**
Let's give this patch another look.  I burned the entire day on it.  Given that we have to turn off unchecked warnings in 2.10 if we don't put it in, let's put it in.  Look at the code, I think it speaks for itself.  I broke it into six little commits for your reviewing pleasure.

### CODE REVIEW

**Comment by paulp** [comment]
Review by @retronym, @jsuereth.

File: src/compiler/scala/tools/nsc/typechecker/Checkable.scala [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  self: Analyzer =>
+
+  import global._
+  import definitions._
+
+  /** The applied type of class 'to' after inferring anything
+   *  possible from the knowledge that 'to' must also be of the
+   *  type given in 'from'.
+   */
+  def propagateKnownTypes(from: Type, to: Symbol): Type = {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Checkable.scala:60]: Modifier patrol: You could make a few defs in here final.

**Comment by retronym** [comment]
Minor comments aside, this looks both rock solid and crystal clear. LGTM.

**Comment by gkossakowski** [comment]
The quality of this pull request is impressive. Well done.

Now the only question is about risk-management. @adriaanm said yesterday that it would be a lot safer to get this into 2.10.1 but I'll try to reach him again (which is hard because he is mostly offline these days) and ask. He might change his mind given the fact that it should be easy to review this change.

**Comment by paulp** [comment]
PLS REBUILD ALL

**Comment by paulp** [comment]
I realized my phrasing up there is still not very clear.  What it should really say is that the value x (of known type X) can be shown not to be a member of P - that we can prove, based on the fact that x.isInstanceOf[X], that !x.isInstanceOf[P].

**Comment by paulp** [comment]
I could really use a concise notation like "<:" for "x is an instance of X." How do people usually write set membership in ascii?

**Comment by paulp** [comment]
Aha, I believe I have nailed the question of impossibility.  Testing now.

**Comment by adriaanm** [comment]
yeah, this `x.isInstanceOf[X], that !x.isInstanceOf[P]` implication also comes up in the pattern matcher analyses, we should factor that out in master

**Comment by paulp** [comment]
PLS REBUILD ALL

**Comment by paulp** [comment]
Okay, I am pretty pleased with this now.  @adriaanm if you're anywhere in the neighborhood I would love to obtain your stamp of approval.  It would give my life meaning.

**Comment by paulp** [comment]
"we should factor that out in master"

I think a lot of this will bear factoring out in master.

**Comment by paulp** [comment]
I'm done polishing this apple in case anyone is in the mood to merge after one more round of tests pass.

**Comment by paulp** [comment]
PLS REBUILD ALL

**Comment by gkossakowski** [comment]
PLS REBUILD pr-scala-testsuite-linux-opt

**Comment by adriaanm** [comment]
LGTM

consider it stamped

**Comment by gkossakowski** [comment]
Cool! Let me just check last time it doesn't break the build.

PLS REBUILD ALL

**Pull request merged**